ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinPushTransitivePredicatesRule.java

```diff
@@ -143,28 +138,82 @@ private ImmutableList<RexNode> getValidPreds(RelOptCluster cluster, RelNode chil
+    // An unsafe predicate in this context is one that when pushed across join operands, can lead
+    // to redundant predicates that cannot be simplified (by means of predicates merging with other existing ones).
+    // This situation can lead to an OOM for cases where lack of simplification allows inferring new predicates
+    // (from LHS to RHS and vice-versa) recursively, predicates which are redundant, but that RexSimplify cannot handle.
+    // This notion can be relaxed as soon as RexSimplify gets more powerful, and it can handle such cases.
+    if (HiveConf.getBoolVar(conf, HiveConf.ConfVars.HIVE_JOIN_PUSH_TRANSITIVE_PREDICATES_CONSERVATIVE)) {
+      toPush = toPush.stream()
+          .filter(unsafeOperatorsFinder::isSafe)
+          .collect(Collectors.toList());
+    }
+
+    return ImmutableList.copyOf(toPush);
   }
 
-  private RexNode getTypeSafePred(RelOptCluster cluster, RexNode rex, RelDataType rType) {
-    RexNode typeSafeRex = rex;
-    if ((typeSafeRex instanceof RexCall) && HiveCalciteUtil.isComparisonOp((RexCall) typeSafeRex)) {
-      RexBuilder rb = cluster.getRexBuilder();
-      List<RexNode> fixedPredElems = new ArrayList<RexNode>();
-      RelDataType commonType = cluster.getTypeFactory().leastRestrictive(
-          RexUtil.types(((RexCall) rex).getOperands()));
-      for (RexNode rn : ((RexCall) rex).getOperands()) {
-        fixedPredElems.add(rb.ensureType(commonType, rn, true));
-      }
+  //~ Inner Classes ----------------------------------------------------------
+
+  /**
+   * Finds unsafe operators in an expression (at any level of nesting).
+   * At the moment, the only unsafe operator is OR.
+   *
+   * Example 1: OR(=($0, $1), IS NOT NULL($2))):INTEGER (OR in the top-level expression)
+   * Example 2: NOT(AND(=($0, $1), IS NOT NULL($2))
+   *   this is equivalent to OR((<>($0, $1), IS NULL($2))
+   * Example 3: AND(OR(=($0, $1), IS NOT NULL($2)))) (OR in inner expression)
+   */
+  private static class UnsafeOperatorsFinder extends RexVisitorImpl<Void> {
+    // accounting for DeMorgan's law
+    boolean inNegation = false;
+    boolean isSafe = true;
+
+    protected UnsafeOperatorsFinder() {
+      super(true);
+    }
 
-      typeSafeRex = rb.makeCall(((RexCall) typeSafeRex).getOperator(), fixedPredElems);
+    @Override
+    public Void visitCall(RexCall call) {
+      switch (call.getKind()) {
+      case OR:
+        if (inNegation) {
+          return super.visitCall(call);
+        } else {
+          this.isSafe = false;
+          return null;
+        }
+      case AND:
+        if (inNegation) {
+          this.isSafe = false;
+          return null;
+        } else {
+          return super.visitCall(call);
+        }
+      case NOT:
+        inNegation = !inNegation;
+        return super.visitCall(call);
+      default:
+        return super.visitCall(call);
+      }
     }
 
-    return typeSafeRex;
+    boolean isSafe(RexNode node) {
```

Comment:
Consider making this method a utility by moving it in `HiveCalciteUtil` e.g., `HiveCalciteUtil#containsDisjunction`. Also maybe it would be safer to make the class disposable, i.e., instantiate only inside the method and then discard. We are going to spend some extra resources for instantiation but I guess the overhead is negligible.